File: golang/tests/pkg/unit/get_cdp_adb_link_test.go
package agentbay_test

import (
	"testing"

	"github.com/alibabacloud-go/tea/dara"
	"github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

// MockClient is a mock implementation of the Client
type MockGetCdpAdbLinkClient struct {
	mock.Mock
}

func (m *MockGetCdpAdbLinkClient) GetCdpLink(request *client.GetCdpLinkRequest) (*client.GetCdpLinkResponse, error) {
	args := m.Called(request)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*client.GetCdpLinkResponse), args.Error(1)
}

func (m *MockGetCdpAdbLinkClient) GetAdbLink(request *client.GetAdbLinkRequest) (*client.GetAdbLinkResponse, error) {
	args := m.Called(request)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*client.GetAdbLinkResponse), args.Error(1)
}

func TestGetCdpLink_Success(t *testing.T) {
	// Arrange
	mockClient := new(MockGetCdpAdbLinkClient)

	expectedURL := "ws://test.cdp.link"
	expectedRequestID := "req123"

	mockResponse := &client.GetCdpLinkResponse{
		Body: &client.GetCdpLinkResponseBody{
			Success:   dara.Bool(true),
			RequestId: dara.String(expectedRequestID),
			Data: &client.GetCdpLinkResponseBodyData{
				Url: dara.String(expectedURL),
			},
		},
	}

	request := &client.GetCdpLinkRequest{
		Authorization: dara.String("Bearer test_token"),
		SessionId:     dara.String("test_session_id"),
	}

	mockClient.On("GetCdpLink", request).Return(mockResponse, nil)

	// Act
	response, err := mockClient.GetCdpLink(request)

	// Assert
	assert.NoError(t, err)
	assert.NotNil(t, response)
	assert.NotNil(t, response.Body)
	assert.True(t, *response.Body.Success)
	assert.Equal(t, expectedURL, *response.Body.Data.Url)
	assert.Equal(t, expectedRequestID, *response.Body.RequestId)

	mockClient.AssertExpectations(t)
}

func TestGetCdpLink_Failure(t *testing.T) {
	// Arrange
	mockClient := new(MockGetCdpAdbLinkClient)

	mockResponse := &client.GetCdpLinkResponse{
		Body: &client.GetCdpLinkResponseBody{
			Success: dara.Bool(false),
			Message: dara.String("API error"),
		},
	}

	request := &client.GetCdpLinkRequest{
		Authorization: dara.String("Bearer test_token"),
		SessionId:     dara.String("invalid_session"),
	}

	mockClient.On("GetCdpLink", request).Return(mockResponse, nil)

	// Act
	response, err := mockClient.GetCdpLink(request)

	// Assert
	assert.NoError(t, err)
	assert.NotNil(t, response)
	assert.NotNil(t, response.Body)
	assert.False(t, *response.Body.Success)
	assert.Equal(t, "API error", *response.Body.Message)

	mockClient.AssertExpectations(t)
}

func TestGetAdbLink_Success(t *testing.T) {
	// Arrange
	mockClient := new(MockGetCdpAdbLinkClient)

	expectedURL := "adb://test.adb.link:5555"
	expectedRequestID := "req456"

	mockResponse := &client.GetAdbLinkResponse{
		Body: &client.GetAdbLinkResponseBody{
			Success:   dara.Bool(true),
			RequestId: dara.String(expectedRequestID),
			Data: &client.GetAdbLinkResponseBodyData{
				Url: dara.String(expectedURL),
			},
		},
	}

	request := &client.GetAdbLinkRequest{
		Authorization: dara.String("Bearer test_token"),
		SessionId:     dara.String("test_session_id"),
		Option:        dara.String(`{"adbkey_pub":"test-public-key"}`),
	}

	mockClient.On("GetAdbLink", request).Return(mockResponse, nil)

	// Act
	response, err := mockClient.GetAdbLink(request)

	// Assert
	assert.NoError(t, err)
	assert.NotNil(t, response)
	assert.NotNil(t, response.Body)
	assert.True(t, *response.Body.Success)
	assert.Equal(t, expectedURL, *response.Body.Data.Url)
	assert.Equal(t, expectedRequestID, *response.Body.RequestId)

	mockClient.AssertExpectations(t)
}

func TestGetAdbLink_Failure(t *testing.T) {
	// Arrange
	mockClient := new(MockGetCdpAdbLinkClient)

	mockResponse := &client.GetAdbLinkResponse{
		Body: &client.GetAdbLinkResponseBody{
			Success: dara.Bool(false),
			Message: dara.String("API error"),
		},
	}

	request := &client.GetAdbLinkRequest{
		Authorization: dara.String("Bearer test_token"),
		SessionId:     dara.String("invalid_session"),
		Option:        dara.String(`{"adbkey_pub":"test-key"}`),
	}

	mockClient.On("GetAdbLink", request).Return(mockResponse, nil)

	// Act
	response, err := mockClient.GetAdbLink(request)

	// Assert
	assert.NoError(t, err)
	assert.NotNil(t, response)
	assert.NotNil(t, response.Body)
	assert.False(t, *response.Body.Success)
	assert.Equal(t, "API error", *response.Body.Message)

	mockClient.AssertExpectations(t)
}

File: typescript/tests/unit/get-cdp-adb-link.test.ts
import { Client } from "../../src/api/client";
import {
  GetCdpLinkRequest,
  GetCdpLinkResponse,
  GetAdbLinkRequest,
  GetAdbLinkResponse,
} from "../../src/api/models/model";
import * as $dara from "@darabonba/typescript";
import { $OpenApiUtil } from "@alicloud/openapi-core";

describe("GetCdpLink and GetAdbLink API", () => {
  let client: Client;
  let mockDoRPCRequest: jest.SpyInstance;

  beforeEach(() => {
    const config = new $OpenApiUtil.Config({
      endpoint: "https://test.endpoint.com",
      accessKeyId: "test-key",
      accessKeySecret: "test-secret",
      regionId: "cn-hangzhou",
    });
    client = new Client(config);
    mockDoRPCRequest = jest.spyOn(client as any, "doRPCRequest");
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe("getCdpLink", () => {
    it("should successfully get CDP link", async () => {
      const mockResponse = {
        headers: {},
        statusCode: 200,
        body: {
          Success: true,
          Data: {
            Url: "ws://test.cdp.link",
          },
          RequestId: "req123",
        },
      };

      mockDoRPCRequest.mockResolvedValue(mockResponse);

      const request = new GetCdpLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
      });

      const response = await client.getCdpLink(request);

      expect(response).toBeInstanceOf(GetCdpLinkResponse);
      expect(response.body?.success).toBe(true);
      expect(response.body?.data?.url).toBe("ws://test.cdp.link");
      expect(response.body?.requestId).toBe("req123");
      expect(mockDoRPCRequest).toHaveBeenCalledTimes(1);
    });

    it("should call getCdpLinkWithOptions with runtime", async () => {
      const mockResponse = {
        headers: {},
        statusCode: 200,
        body: {
          Success: true,
          Data: {
            Url: "ws://test.cdp.link",
          },
        },
      };

      mockDoRPCRequest.mockResolvedValue(mockResponse);

      const request = new GetCdpLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
      });
      const runtime = new $dara.RuntimeOptions({});

      const response = await client.getCdpLinkWithOptions(request, runtime);

      expect(response).toBeInstanceOf(GetCdpLinkResponse);
      expect(response.body?.success).toBe(true);
      expect(mockDoRPCRequest).toHaveBeenCalledTimes(1);
    });

    it("should handle getCdpLink failure", async () => {
      mockDoRPCRequest.mockRejectedValue(new Error("API call failed"));

      const request = new GetCdpLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
      });

      await expect(client.getCdpLink(request)).rejects.toThrow(
        "API call failed"
      );
    });
  });

  describe("getAdbLink", () => {
    it("should successfully get ADB link", async () => {
      const mockResponse = {
        headers: {},
        statusCode: 200,
        body: {
          Success: true,
          Data: {
            Url: "adb://test.adb.link:5555",
          },
          RequestId: "req456",
        },
      };

      mockDoRPCRequest.mockResolvedValue(mockResponse);

      const request = new GetAdbLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
        option: JSON.stringify({ adbkey_pub: "test-public-key" }),
      });

      const response = await client.getAdbLink(request);

      expect(response).toBeInstanceOf(GetAdbLinkResponse);
      expect(response.body?.success).toBe(true);
      expect(response.body?.data?.url).toBe("adb://test.adb.link:5555");
      expect(response.body?.requestId).toBe("req456");
      expect(mockDoRPCRequest).toHaveBeenCalledTimes(1);
    });

    it("should call getAdbLinkWithOptions with runtime", async () => {
      const mockResponse = {
        headers: {},
        statusCode: 200,
        body: {
          Success: true,
          Data: {
            Url: "adb://test.adb.link:5555",
          },
        },
      };

      mockDoRPCRequest.mockResolvedValue(mockResponse);

      const request = new GetAdbLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
        option: JSON.stringify({ adbkey_pub: "test-public-key" }),
      });
      const runtime = new $dara.RuntimeOptions({});

      const response = await client.getAdbLinkWithOptions(request, runtime);

      expect(response).toBeInstanceOf(GetAdbLinkResponse);
      expect(response.body?.success).toBe(true);
      expect(mockDoRPCRequest).toHaveBeenCalledTimes(1);
    });

    it("should handle getAdbLink failure", async () => {
      mockDoRPCRequest.mockRejectedValue(new Error("API call failed"));

      const request = new GetAdbLinkRequest({
        authorization: "Bearer test_token",
        sessionId: "test_session_id",
        option: JSON.stringify({ adbkey_pub: "test-key" }),
      });

      await expect(client.getAdbLink(request)).rejects.toThrow(
        "API call failed"
      );
    });
  });
});

